### PR TITLE
feature(lxy): add gumbel muzero

### PR DESCRIPTION
**Pull request by karroyan**

### CODE REVIEW

File: lzero/policy/gumbel_muzero.py [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+
+    # @profile
+    def _forward_learn(self, data: ttorch.Tensor) -> Dict[str, Union[float, int]]:
+        self._learn_model.train()
+        self._target_model.train()
+
+        # TODO(pu): priority
+        current_batch, target_batch = data
+        obs_batch_ori, action_batch, new_policy_batch, mask_batch, indices, weights, make_time = current_batch
+        target_reward, target_value, target_policy = target_batch
```
Comment (puyuan1996) [inline lzero/policy/gumbel_muzero.py:231]: 这里的target_value已经是gumbel里面的mixed value了吗？

File: lzero/worker/gumbel_muzero_collector.py [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+        # put the game block into the pool
+        self.game_segment_pool.append((last_game_segments[i], last_game_priorities[i], done[i]))
+
+        # reset last game_segments
+        last_game_segments[i] = None
+        last_game_priorities[i] = None
+
+        return None
+
+    def collect(self,
```
Comment (puyuan1996) [inline lzero/worker/gumbel_muzero_collector.py:266]: 同样的，与muzero_collector有哪些不同呢？可以继承MuZeroCollector，然后改动用注释重点标记一下

File: lzero/mcts/buffer/game_buffer_gumbel_muzero.py [hunk 3]
```diff
@@ -0,0 +1,704 @@
...
+from lzero.mcts.utils import prepare_observation
+from lzero.policy import to_detach_cpu_numpy, concat_output, concat_output_value, inverse_scalar_transform
+from .game_buffer import GameBuffer
+
+if TYPE_CHECKING:
+    from lzero.policy import MuZeroPolicy, EfficientZeroPolicy, SampledEfficientZeroPolicy
+
+
+@BUFFER_REGISTRY.register('game_buffer_gumbel_muzero')
+class GumbelMuZeroGameBuffer(GameBuffer):
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply lzero/mcts/buffer/game_buffer_gumbel_muzero.py:18]: buffer里面准备的targe tvalue需要是mixed value吗？

File: lzero/policy/gumbel_muzero.py [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+
+
+    # @profile
+    def _forward_learn(self, data: ttorch.Tensor) -> Dict[str, Union[float, int]]:
+        self._learn_model.train()
+        self._target_model.train()
+
+        # TODO(pu): priority
+        current_batch, target_batch = data
+        obs_batch_ori, action_batch, new_policy_batch, mask_batch, indices, weights, make_time = current_batch
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply lzero/policy/gumbel_muzero.py:230]: target_policy 和 new_policy_batch 的含义和作用分别是？在buffer中是怎么产生的呢？

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: lzero/mcts/buffer/game_buffer_gumbel_muzero.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (karroyan) [reply lzero/mcts/buffer/game_buffer_gumbel_muzero.py:18]: > buffer里面准备的targe tvalue需要是mixed value吗？

不需要。实际上gumbel muzero没有用到这里的target value，而是用mcts返回的improved policy作为更新目标

File: lzero/mcts/ctree/ctree_gumbel_muzero/gmz_tree.pyx [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+def softmax(list num_list):
+    cdef vector[float] cnum_list = num_list;
+    cdef int clength = len(num_list)
+    csoftmax(cnum_list, clength)
+    for i in range(len(num_list)):
+        num_list[i] = cnum_list[i]
+    return num_list
+
+def pcompute_mixed_value(float raw_value, list q_values, list child_visit, list child_prior):
```
[the remark this answers is not in the document]
Comment (karroyan) [reply lzero/mcts/ctree/ctree_gumbel_muzero/gmz_tree.pyx:119]: done，只修改了gumbel muzero增加的函数，原有函数还是和muzero保持一致了

File: lzero/policy/gumbel_muzero.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (karroyan) [reply lzero/policy/gumbel_muzero.py]: target_policy还是muzero中的target policy，improved_policy_batch(原new_policy_batch)是gumbel muzero中的improved policy /pi'=softmax(logits+\sigma(completedQ))。improved_policy_batch是在collect data时从mcts中收集的：https://github.com/karroyan/LightZero/blob/64b5cde7bcd17bf73967990a58c018e5bfbec01d/lzero/policy/gumbel_muzero.py#L517

File: lzero/mcts/buffer/game_segment.py [hunk 6]
```diff
@@ -181,7 +183,7 @@ def get_targets(self, timestep: int) -> Tuple:
         return self.target_values[timestep], self.target_rewards[timestep], self.target_policies[timestep]
 
     def store_search_stats(
-            self, visit_counts: List, root_value: List, root_sampled_actions=None, idx: int = None
+            self, visit_counts: List, root_value: List, root_sampled_actions=None, improved_policy=None, idx: int = None
```
Comment (PaParaZz1) [inline lzero/mcts/buffer/game_segment.py:186]: typing lint for `root_sampled_actions` and `improved_policy`

File: zoo/board_games/tictactoe/config/tictactoe_gumbel_muzero_bot_mode_config.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        n_evaluator_episode=evaluator_env_num,
+        manager=dict(shared_memory=False, ),
+    ),
+    policy=dict(
+        model=dict(
+            observation_shape=(3, 3, 3),
+            action_space_size=9,
+            image_channel=3,
+            # We use the small size model for tictactoe.
+            num_res_blocks=1,
```
Comment (PaParaZz1) [inline zoo/board_games/tictactoe/config/tictactoe_gumbel_muzero_bot_mode_config.py:34]: simplify config, remove unchanged fields

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: lzero/mcts/tree_search/mcts_ctree.py [hunk 8]
```diff
@@ -351,3 +353,120 @@ def search(
...
+                    latent_states.append(latent_state_pool[ix][iy])
+
+                latent_states = torch.from_numpy(np.asarray(latent_states)).to(device).float()
+                last_actions = torch.from_numpy(np.asarray(last_actions)).to(device).unsqueeze(1).long()
+
+                # evaluation for leaf nodes
+                network_output = model.recurrent_inference(latent_states, last_actions)
+                # TODO(pu)
+                if not model.training:
+                    # if not in training, obtain the scalars of the value/reward
```
Comment (puyuan1996) [inline lzero/mcts/tree_search/mcts_ctree.py:452]: 仿照这里 https://github.com/opendilab/LightZero/blob/main/lzero/mcts/tree_search/mcts_ctree.py#L324 更新文件里面的变量名，注释以及一些方法哈

File: lzero/policy/gumbel_muzero.py [hunk 9]
```diff
@@ -0,0 +1,693 @@
+import copy
+from typing import List, Dict, Any, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+import torch.optim as optim
+import treetensor.torch as ttorch
```
Comment (puyuan1996) [inline lzero/policy/gumbel_muzero.py:8]: 参考https://github.com/opendilab/LightZero/blob/main/lzero/policy/muzero.py里面的还是用torch哈

File: lzero/policy/gumbel_muzero.py [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+from ding.model import model_wrap
+from ding.policy.base_policy import Policy
+from ding.rl_utils import get_nstep_return_data, get_train_sample
+from ding.torch_utils import to_tensor, to_device
+from ding.utils import POLICY_REGISTRY
+from torch.nn import L1Loss, KLDivLoss
+
+# python mcts
```
Comment (puyuan1996) [inline lzero/policy/gumbel_muzero.py:16]: 更新注释

File: lzero/policy/gumbel_muzero.py [hunk 11]
```diff
@@ -0,0 +1,693 @@
...
+from lzero.policy import scalar_transform, InverseScalarTransform, cross_entropy_loss, phi_transform, \
+    DiscreteSupport, to_torch_float_tensor, mz_network_output_unpack, select_action, negative_cosine_similarity, prepare_obs, \
+    configure_optimizers
+
+
+@POLICY_REGISTRY.register('gumbel_muzero')
+class GumeblMuZeroPolicy(Policy):
+    """
+    Overview:
+        The policy class for Gumbel Muzero
```
Comment (puyuan1996) [inline lzero/policy/gumbel_muzero.py:29]: 末尾加上句号

File: lzero/policy/gumbel_muzero.py [hunk 12]
```diff
@@ -0,0 +1,693 @@
...
+        # (int) How many updates(iterations) to train after collector's one collection.
+        # Bigger "update_per_collect" means bigger off-policy.
+        # collect data -> update policy-> collect data -> ...
+        # For different env, we have different episode_length,
+        # we usually set update_per_collect = collector_env_num * episode_length / batch_size * reuse_factor
+        update_per_collect=100,
+        # (int) Minibatch size for one gradient descent.
+        batch_size=256,
+        # (str) Optimizer for training policy network. ['SGD' or 'Adam']
+        optim_type='AdamW',
```
Comment (puyuan1996) [inline lzero/policy/gumbel_muzero.py:99]: 默认值设置为何muzero一样哈https://github.com/opendilab/LightZero/blob/main/lzero/policy/muzero.py#L101

File: lzero/worker/gumbel_muzero_collector.py [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+            search_values = torch.from_numpy(np.array(search_values_lst[i])).to(self.policy_config.device
+                                                                                ).float().view(-1)
+            # ==============================================================
+            # Gumbel Muzero related code
+            # ==============================================================
+            improved_policy = torch.from_numpy(np.array(improved_policy_lst[i])).to(self.policy_config.device
+                                                                                ).float().view(-1)
+            priorities = L1Loss(reduction='none'
+                                )(pred_values,
+                                  search_values, improved_policy).detach().cpu().numpy() + self.policy_config.prioritized_replay_eps
```
Comment (puyuan1996) [inline lzero/worker/gumbel_muzero_collector.py:54]: 这里的 priorities 计算为什么传入了3个参数呀？priorities的计算似乎有问题哈

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: lzero/worker/gumbel_muzero_collector.py (hunk 13 above)
Comment (karroyan) [reply lzero/worker/gumbel_muzero_collector.py]: 已修改

File: lzero/policy/gumbel_muzero.py [hunk 14]
```diff
@@ -0,0 +1,700 @@
...
+                # NOTE: Only legal actions possess visit counts, so the ``action_index_in_legal_action_set`` represents
+                # the index within the legal action set, rather than the index in the entire action set.
+                #  Setting deterministic=True implies choosing the action with the highest value (argmax) rather than
+                # sampling during the evaluation phase.
+                action_index_in_legal_action_set, visit_count_distribution_entropy = select_action(
+                    distributions, temperature=1, deterministic=True
+                )
+                # NOTE: Convert the ``action_index_in_legal_action_set`` to the corresponding ``action`` in the
+                # entire action set.
+                action = np.where(action_mask[i] == 1.0)[0][action_index_in_legal_action_set]
```
Comment (puyuan1996) [inline lzero/policy/gumbel_muzero.py:631]: eval还是根据的visit_count_distributions选择的argmax?

File: lzero/mcts/buffer/game_buffer_gumbel_muzero.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+        context = reward_value_context, policy_re_context, policy_non_re_context, current_batch
+        return context
+
+    def update_priority(self, train_data: List[np.ndarray], batch_priorities: Any) -> None:
+        """
+        Overview:
+            Update the priority of training data.
+        Arguments:
+            - train_data (:obj:`Optional[List[Optional[np.ndarray]]]`): training data to be updated priority.
```
Comment (PaParaZz1) [inline lzero/mcts/buffer/game_buffer_gumbel_muzero.py:130]: why `Optional` here

File: lzero/mcts/buffer/game_buffer_gumbel_muzero.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+            - batch_priorities (:obj:`batch_priorities`): priorities to update to.
+        NOTE:
+            train_data = [current_batch, target_batch]
+            current_batch = [obs_list, action_list, mask_list, batch_index_list, weights, make_time_list]
+        """
+        indices = train_data[0][4]
+        metas = {'make_time': train_data[0][6], 'batch_priorities': batch_priorities}
```
Comment (PaParaZz1) [inline lzero/mcts/buffer/game_buffer_gumbel_muzero.py:137]: maybe we can rewrite it to `train_data[0][-1]` to let muzero and gumbel use the same function

File: lzero/mcts/buffer/game_segment.py [hunk 17]
```diff
@@ -141,7 +143,7 @@ def append(
 
     def pad_over(
             self, next_segment_observations: List, next_segment_rewards: List, next_segment_root_values: List,
-            next_segment_child_visits: List
+            next_segment_child_visits: List, next_segment_improved_policy: List = None
```
Comment (PaParaZz1) [inline lzero/mcts/buffer/game_segment.py:146]: add comments for the new argument

File: lzero/mcts/tree_search/mcts_ctree.py [hunk 18]
```diff
@@ -351,3 +353,120 @@ def search(
                     current_latent_state_index, discount_factor, reward_batch, value_batch, policy_logits_batch,
                     min_max_stats_lst, results, virtual_to_play_batch
                 )
+
+class GumbelMuZeroMCTSCtree(object):
```
Comment (PaParaZz1) [inline lzero/mcts/tree_search/mcts_ctree.py:357]: add unittest for this class

File: lzero/policy/gumbel_muzero.py [hunk 19]
```diff
@@ -0,0 +1,700 @@
...
+from lzero.policy import scalar_transform, InverseScalarTransform, cross_entropy_loss, phi_transform, \
+    DiscreteSupport, to_torch_float_tensor, mz_network_output_unpack, select_action, negative_cosine_similarity, prepare_obs, \
+    configure_optimizers
+
+
+@POLICY_REGISTRY.register('gumbel_muzero')
+class GumeblMuZeroPolicy(Policy):
+    """
+    Overview:
+        The policy class for Gumbel Muzero
```
Comment (PaParaZz1) [inline lzero/policy/gumbel_muzero.py:29]: add paper link

File: lzero/policy/gumbel_muzero.py [hunk 20]
```diff
@@ -0,0 +1,700 @@
...
+        # packing loss info for tensorboard logging
+        loss_info = (
+            weighted_total_loss.item(), loss.mean().item(), policy_loss.mean().item(), reward_loss.mean().item(),
+            value_loss.mean().item(), consistency_loss.mean(), entropy_loss.mean()
+        )
+        if self._cfg.monitor_extra_statistics:
+            predicted_rewards = torch.stack(predicted_rewards).transpose(1, 0).squeeze(-1)
+            predicted_rewards = predicted_rewards.reshape(-1).unsqueeze(-1)
+
+            td_data = (
```
Comment (PaParaZz1) [inline lzero/policy/gumbel_muzero.py:408]: why to construct a variable named `td_data`, just use `xxx.mean.item()` and we can the value to be logged

File: lzero/policy/gumbel_muzero.py [hunk 21]
```diff
@@ -0,0 +1,700 @@
...
+
+        gradient_scale = 1 / self._cfg.num_unroll_steps
+        weighted_total_loss.register_hook(lambda grad: grad * gradient_scale)
+        self._optimizer.zero_grad()
+        weighted_total_loss.backward()
+        total_grad_norm_before_clip = torch.nn.utils.clip_grad_norm_(
+            self._learn_model.parameters(), self._cfg.grad_clip_value
+        )
+        self._optimizer.step()
+        if self._cfg.lr_piecewise_constant_decay is True:
```
Comment (PaParaZz1) [inline lzero/policy/gumbel_muzero.py:391]: if self._cfg.lr_piecewise_constant_decay

File: lzero/policy/gumbel_muzero.py [hunk 22]
```diff
@@ -0,0 +1,700 @@
...
+            obs_batch = self.image_transforms.transform(obs_batch)
+            if self._cfg.model.self_supervised_learning_loss:
+                obs_target_batch = self.image_transforms.transform(obs_target_batch)
+
+        # shape: (batch_size, num_unroll_steps, action_dim)
+        # NOTE: .long(), in discrete action space.
+        action_batch = torch.from_numpy(action_batch).to(self._cfg.device).unsqueeze(-1).long()
+        data_list = [
+            mask_batch,
+            target_reward.astype('float64'),
```
Comment (PaParaZz1) [inline lzero/policy/gumbel_muzero.py:248]: why float64 rather than float32

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: zoo/atari/config/atari_gumbel_muzero_config.py [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+            self_supervised_learning_loss=True,  # default is False
+        ),
+        cuda=True,
+        env_type='not_board_games',
+        game_segment_length=400,
+        use_augmentation=True,
+        update_per_collect=update_per_collect,
+        batch_size=batch_size,
+        optim_type='SGD',
+        max_num_considered_actions=3,
```
Comment (puyuan1996) [inline zoo/atari/config/atari_gumbel_muzero_config.py:59]: max_num_considered_actions是否需要设置为6？

File: lzero/policy/gumbel_muzero.py (hunk 1 above)
Comment (karroyan) [reply lzero/policy/gumbel_muzero.py:233]: 嗯是的

File: lzero/policy/gumbel_muzero.py (hunk 14 above)
Comment (karroyan) [reply lzero/policy/gumbel_muzero.py:631]: 已改为improved_policy

File: lzero/entry/train_muzero.py [hunk 24]
```diff
@@ -43,15 +42,19 @@ def train_muzero(
...
     if create_cfg.policy.type == 'muzero':
         from lzero.mcts import MuZeroGameBuffer as GameBuffer
     elif create_cfg.policy.type == 'efficientzero':
         from lzero.mcts import EfficientZeroGameBuffer as GameBuffer
     elif create_cfg.policy.type == 'sampled_efficientzero':
         from lzero.mcts import SampledEfficientZeroGameBuffer as GameBuffer
+    elif create_cfg.policy.type == 'gumbel_muzero':
+        from lzero.mcts import GumbelMuZeroGameBuffer as GameBuffer
+    from lzero.worker import MuZeroCollector as Collector
+    from lzero.worker import MuZeroEvaluator as Evaluator
```
Comment (PaParaZz1) [inline lzero/entry/train_muzero.py:57]: move this import to the top part of this file

File: lzero/mcts/tree_search/mcts_ctree.py [hunk 25]
```diff
@@ -321,19 +323,17 @@ def search(
                 """
                 network_output = model.recurrent_inference(latent_states, last_actions)
 
-                if not model.training:
-                    # if not in training, obtain the scalars of the value/reward
+                [
```
Comment (PaParaZz1) [inline lzero/mcts/tree_search/mcts_ctree.py:326]: it is quite ugly here, maybe you should separate them to four lines

File: lzero/mcts/tree_search/mcts_ctree.py [hunk 26]
```diff
@@ -351,3 +351,149 @@ def search(
...
+        from lzero.mcts.ctree.ctree_gumbel_muzero import gmz_tree as ctree
+        return ctree.Roots(active_collect_env_num, legal_actions)
+
+    def search(self, roots: Any, model: torch.nn.Module, latent_state_roots: List[Any], to_play_batch: Union[int,
+                                                                                                          List[Any]]
+    ) -> None:
+        """
+        Overview:
+            Do MCTS for the roots (a batch of root nodes in parallel). Parallel in model inference.
+             Use the cpp tree.
```
Comment (PaParaZz1) [inline lzero/mcts/tree_search/mcts_ctree.py:412]: polish the indent

File: lzero/worker/muzero_collector.py [hunk 27]
```diff
@@ -441,13 +465,17 @@ def collect(self,
...
+                        completed_value_lst[env_id] += np.mean(np.array(completed_value_dict[env_id]))
 
                     eps_steps_lst[env_id] += 1
                     total_transitions += 1
 
                     if self.policy_config.use_priority and not self.policy_config.use_max_priority_for_new_data:
                         pred_values_lst[env_id].append(pred_value_dict[env_id])
                         search_values_lst[env_id].append(value_dict[env_id])
+                        if self.policy_config.gumbel_algo:
+                                improved_policy_lst[env_id].append(improved_policy_dict[env_id])
```
Comment (PaParaZz1) [inline lzero/worker/muzero_collector.py:478]: polish the indent

File: lzero/worker/muzero_collector.py [hunk 28]
```diff
@@ -490,12 +520,21 @@ def collect(self,
                 if timestep.done:
                     self._total_episode_count += 1
                     reward = timestep.info['eval_episode_return']
-                    info = {
-                        'reward': reward,
-                        'time': self._env_info[env_id]['time'],
-                        'step': self._env_info[env_id]['step'],
-                        'visit_entropy': visit_entropies_lst[env_id] / eps_steps_lst[env_id],
-                    }
+                    if self.policy_config.gumbel_algo:
```
Comment (PaParaZz1) [inline lzero/worker/muzero_collector.py:523]: first create four common variables, then append the one related to gumbel muzero

File: lzero/worker/muzero_collector.py [hunk 29]
```diff
@@ -615,24 +654,46 @@ def _output_log(self, train_iter: int) -> None:
...
-                'reward_std': np.std(episode_reward),
-                'reward_max': np.max(episode_reward),
-                'reward_min': np.min(episode_reward),
-                'total_envstep_count': self._total_envstep_count,
-                'total_episode_count': self._total_episode_count,
-                'total_duration': self._total_duration,
-                'visit_entropy': np.mean(visit_entropy),
-                # 'each_reward': episode_reward,
-            }
+            if self.policy_config.gumbel_algo:
```
Comment (PaParaZz1) [inline lzero/worker/muzero_collector.py:660]: polish here like what I comment above

File: zoo/atari/config/atari_muzero_config.py [hunk 30]
```diff
@@ -1,7 +1,7 @@
 from easydict import EasyDict
 
 # options={'PongNoFrameskip-v4', 'QbertNoFrameskip-v4', 'MsPacmanNoFrameskip-v4', 'SpaceInvadersNoFrameskip-v4', 'BreakoutNoFrameskip-v4', ...}
-env_name = 'PongNoFrameskip-v4'
+env_name = 'MsPacmanNoFrameskip-v4'
```
Comment (PaParaZz1) [inline zoo/atari/config/atari_muzero_config.py:4]: why modify this

File: zoo/box2d/lunarlander/config/lunarlander_disc_muzero_config.py [hunk 31]
```diff
@@ -6,7 +6,7 @@
 collector_env_num = 8
 n_episode = 8
 evaluator_env_num = 3
-num_simulations = 50
+num_simulations = 5
```
Comment (PaParaZz1) [inline zoo/box2d/lunarlander/config/lunarlander_disc_muzero_config.py:9]: why modify this

File: zoo/board_games/tictactoe/config/tictactoe_gumbel_muzero_bot_mode_config.py [hunk 32]
```diff
@@ -0,0 +1,90 @@
...
+    exp_name=
+    f'data_mz_ctree/tictactoe_gumbel_muzero_bot-mode_ns{num_simulations}_upc{update_per_collect}_rr{reanalyze_ratio}_seed0',
+    env=dict(
+        battle_mode='play_with_bot_mode',
+        collector_env_num=collector_env_num,
+        evaluator_env_num=evaluator_env_num,
+        n_evaluator_episode=evaluator_env_num,
+        manager=dict(shared_memory=False, ),
+    ),
+    policy=dict(
```
Comment (PaParaZz1) [inline zoo/board_games/tictactoe/config/tictactoe_gumbel_muzero_bot_mode_config.py:28]: can we simplify some fields with its default value

File: lzero/policy/gumbel_muzero.py [hunk 33]
```diff
@@ -0,0 +1,745 @@
...
+        """
+        self._collect_model.eval()
+        self.collect_mcts_temperature = temperature
+        active_collect_env_num = data.shape[0]
+        with torch.no_grad():
+            # data shape [B, S x C, W, H], e.g. {Tensor:(B, 12, 96, 96)}
+            network_output = self._collect_model.initial_inference(data)
+            latent_state_roots, reward_roots, pred_values, policy_logits = mz_network_output_unpack(network_output)
+
+            if not self._learn_model.training:
```
Comment (PaParaZz1) [inline lzero/policy/gumbel_muzero.py:524]: why this if branch here

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: zoo/board_games/tictactoe/config/tictactoe_gumbel_muzero_bot_mode_config.py (hunk 32 above)
Comment (karroyan) [reply zoo/board_games/tictactoe/config/tictactoe_gumbel_muzero_bot_mode_config.py:28]: all the configs that are not frequently changed here differ from the default config

**Pull request merged**